File: bindings/go/stores/redis/redis.go
package redis

import (
	"bytes"
	"context"
	"encoding/binary"
	"errors"
	"fmt"
	"sort"
	"strings"

	prolly "build.crab/prolly-go"
	redisclient "github.com/redis/go-redis/v9"
)

var rootCASScript = redisclient.NewScript(`
local current = redis.call('GET', KEYS[1])
if ARGV[1] == '1' then
  if current == false or current ~= ARGV[2] then return {0, current} end
else
  if current ~= false then return {0, current} end
end
if ARGV[3] == '1' then redis.call('SET', KEYS[1], ARGV[4]) else redis.call('DEL', KEYS[1]) end
return {1, false}
`)

var transactionScript = redisclient.NewScript(`
local condition_count = tonumber(ARGV[1])
local node_write_count = tonumber(ARGV[2])
local root_write_count = tonumber(ARGV[3])
local arg_index = 4
for i = 1, condition_count do
  local current = redis.call('GET', KEYS[i])
  local has_expected = ARGV[arg_index]
  local expected = ARGV[arg_index + 1]
  arg_index = arg_index + 2
  if has_expected == '1' then
    if current == false or current ~= expected then return {0, i, current} end
  else
    if current ~= false then return {0, i, current} end
  end
end
local node_offset = condition_count
for i = 1, node_write_count do
  local kind = ARGV[arg_index]
  arg_index = arg_index + 1
  local key = KEYS[node_offset + i]
  if kind == 'upsert' then redis.call('SET', key, ARGV[arg_index]); arg_index = arg_index + 1
  elseif kind == 'delete' then redis.call('DEL', key)
  else error('unknown transaction node op') end
end
local root_offset = condition_count + node_write_count
for i = 1, root_write_count do
  local kind = ARGV[arg_index]
  arg_index = arg_index + 1
  local key = KEYS[root_offset + i]
  if kind == 'put' then redis.call('SET', key, ARGV[arg_index]); arg_index = arg_index + 1
  elseif kind == 'delete' then redis.call('DEL', key)
  else error('unknown transaction root op') end
end
return {1, 0, false}
`)

type Options struct {
	AdapterName     string
	KeyPrefix       []byte
	ReadParallelism uint32
}

type Store struct {
	client  redisclient.UniversalClient
	options Options
}

func New(client redisclient.UniversalClient, options Options) *Store {
	if strings.TrimSpace(options.AdapterName) == "" {
		options.AdapterName = "redis-v1"
	}
	if options.KeyPrefix == nil {
		options.KeyPrefix = []byte("prolly:")
	}
	options.KeyPrefix = clone(options.KeyPrefix)
	if options.ReadParallelism == 0 {
		options.ReadParallelism = 16
	}
	return &Store{client: client, options: options}
}

func (s *Store) Descriptor(ctx context.Context) (prolly.StoreDescriptor, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.StoreDescriptor{}, err
	}
	return prolly.StoreDescriptor{
		ProtocolMajor: 1, AdapterName: s.options.AdapterName, Provider: "redis", SchemaVersion: 1,
		Capabilities: prolly.StoreCapabilities{
			NativeBatchReads: true, AtomicBatchWrites: true, NodeScan: true, Hints: true,
			AtomicNodesAndHint: true, RootScan: true, RootCompareAndSwap: true,
			Transactions: true, ReadParallelism: s.options.ReadParallelism,
		},
	}, nil
}

func (s *Store) GetNode(ctx context.Context, key []byte) (prolly.OptionalBytes, error) {
	return s.get(ctx, s.nodeKey(key))
}
func (s *Store) PutNode(ctx context.Context, key, value []byte) error {
	return redisError("put_node", s.client.Set(ctx, s.nodeKey(key), value, 0).Err())
}
func (s *Store) DeleteNode(ctx context.Context, key []byte) error {
	return redisError("delete_node", s.client.Del(ctx, s.nodeKey(key)).Err())
}

func (s *Store) BatchNodes(ctx context.Context, mutations []prolly.NodeMutation) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	_, err := s.client.TxPipelined(ctx, func(pipe redisclient.Pipeliner) error {
		for _, mutation := range mutations {
			if mutation.Value.Present {
				pipe.Set(ctx, s.nodeKey(mutation.Key), mutation.Value.Value, 0)
			} else {
				pipe.Del(ctx, s.nodeKey(mutation.Key))
			}
		}
		return nil
	})
	return redisError("batch_nodes", err)
}

func (s *Store) BatchGetNodesOrdered(ctx context.Context, keys [][]byte) ([]prolly.OptionalBytes, error) {
	if err := s.ready(ctx); err != nil {
		return nil, err
	}
	redisKeys := make([]string, len(keys))
	for index, key := range keys {
		redisKeys[index] = s.nodeKey(key)
	}
	values, err := s.client.MGet(ctx, redisKeys...).Result()
	if err != nil {
		return nil, redisError("batch_get", err)
	}
	result := make([]prolly.OptionalBytes, len(values))
	for index, value := range values {
		if value == nil {
			continue
		}
		switch typed := value.(type) {
		case string:
			result[index] = prolly.PresentBytes([]byte(typed))
		case []byte:
			result[index] = prolly.PresentBytes(typed)
		default:
			return nil, &prolly.StoreError{Code: "invalid_result", Message: fmt.Sprintf("unexpected Redis MGET value %T", value)}
		}
	}
	return result, nil
}

func (s *Store) ListNodeCIDs(ctx context.Context) ([][]byte, error) {
	keys, err := s.scanFamily(ctx, []byte("node:"))
	if err != nil {
		return nil, err
	}
	result := make([][]byte, 0, len(keys))
	for _, key := range keys {
		if len(key) == 32 {
			result = append(result, key)
		}
	}
	sort.Slice(result, func(i, j int) bool { return bytes.Compare(result[i], result[j]) < 0 })
	return result, nil
}

func (s *Store) GetHint(ctx context.Context, namespace, key []byte) (prolly.OptionalBytes, error) {
	return s.get(ctx, s.hintKey(namespace, key))
}
func (s *Store) PutHint(ctx context.Context, namespace, key, value []byte) error {
	return redisError("put_hint", s.client.Set(ctx, s.hintKey(namespace, key), value, 0).Err())
}
func (s *Store) BatchPutNodesWithHint(ctx context.Context, nodes []prolly.NodeEntry, namespace, key, value []byte) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	_, err := s.client.TxPipelined(ctx, func(pipe redisclient.Pipeliner) error {
		for _, node := range nodes {
			pipe.Set(ctx, s.nodeKey(node.Key), node.Value, 0)
		}
		pipe.Set(ctx, s.hintKey(namespace, key), value, 0)
		return nil
	})
	return redisError("batch_nodes_hint", err)
}

func (s *Store) GetRootManifest(ctx context.Context, name []byte) (prolly.OptionalBytes, error) {
	return s.get(ctx, s.rootKey(name))
}
func (s *Store) PutRootManifest(ctx context.Context, name, manifest []byte) error {
	return redisError("put_root", s.client.Set(ctx, s.rootKey(name), manifest, 0).Err())
}
func (s *Store) DeleteRootManifest(ctx context.Context, name []byte) error {
	return redisError("delete_root", s.client.Del(ctx, s.rootKey(name)).Err())
}

func (s *Store) CompareAndSwapRootManifest(ctx context.Context, name []byte, expected, replacement prolly.OptionalBytes) (prolly.RootCASResult, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.RootCASResult{}, err
	}
	response, err := rootCASScript.Run(ctx, s.client, []string{s.rootKey(name)}, boolString(expected.Present), expected.Value, boolString(replacement.Present), replacement.Value).Result()
	if err != nil {
		return prolly.RootCASResult{}, redisError("root_cas", err)
	}
	values, ok := response.([]interface{})
	if !ok || len(values) != 2 {
		return prolly.RootCASResult{}, invalidRedisResult("root CAS", response)
	}
	applied, err := redisBool(values[0])
	if err != nil {
		return prolly.RootCASResult{}, err
	}
	if applied {
		return prolly.RootCASResult{Applied: true, Current: replacement.Clone()}, nil
	}
	current, err := redisOptional(values[1])
	return prolly.RootCASResult{Current: current}, err
}

func (s *Store) ListRootManifests(ctx context.Context) ([]prolly.NamedStoreRoot, error) {
	names, err := s.scanFamily(ctx, []byte("root:"))
	if err != nil {
		return nil, err
	}
	sort.Slice(names, func(i, j int) bool { return bytes.Compare(names[i], names[j]) < 0 })
	result := make([]prolly.NamedStoreRoot, 0, len(names))
	for _, name := range names {
		value, err := s.GetRootManifest(ctx, name)
		if err != nil {
			return nil, err
		}
		if value.Present {
			result = append(result, prolly.NamedStoreRoot{Name: name, Manifest: value.Value})
		}
	}
	return result, nil
}

func (s *Store) CommitTransaction(ctx context.Context, nodes []prolly.NodeMutation, conditions []prolly.RootCondition, roots []prolly.RootWrite) (prolly.StoreTransactionResult, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.StoreTransactionResult{}, err
	}
	keys := make([]string, 0, len(conditions)+len(nodes)+len(roots))
	args := []interface{}{len(conditions), len(nodes), len(roots)}
	for _, condition := range conditions {
		keys = append(keys, s.rootKey(condition.Name))
		args = append(args, boolString(condition.Expected.Present), condition.Expected.Value)
	}
	for _, node := range nodes {
		keys = append(keys, s.nodeKey(node.Key))
		if node.Value.Present {
			args = append(args, "upsert", node.Value.Value)
		} else {
			args = append(args, "delete")
		}
	}
	for _, root := range roots {
		keys = append(keys, s.rootKey(root.Name))
		if root.Replacement.Present {
			args = append(args, "put", root.Replacement.Value)
		} else {
			args = append(args, "delete")
		}
	}
	response, err := transactionScript.Run(ctx, s.client, keys, args...).Result()
	if err != nil {
		return prolly.StoreTransactionResult{}, redisError("transaction", err)
	}
	values, ok := response.([]interface{})
	if !ok || len(values) != 3 {
		return prolly.StoreTransactionResult{}, invalidRedisResult("transaction", response)
	}
	applied, err := redisBool(values[0])
	if err != nil {
		return prolly.StoreTransactionResult{}, err
	}
	if applied {
		return prolly.StoreTransactionResult{Applied: true}, nil
	}
	index, ok := values[1].(int64)
	if !ok || index < 1 || int(index) > len(conditions) {
		return prolly.StoreTransactionResult{}, invalidRedisResult("transaction conflict index", values[1])
	}
	current, err := redisOptional(values[2])
	if err != nil {
		return prolly.StoreTransactionResult{}, err
	}
	condition := conditions[index-1]
	return prolly.StoreTransactionResult{Conflict: &prolly.StoreTransactionConflict{Name: clone(condition.Name), Expected: condition.Expected.Clone(), Current: current}}, nil
}

func (s *Store) Clear(ctx context.Context) error {
	if len(s.options.KeyPrefix) == 0 {
		return &prolly.StoreError{Code: "invalid_argument", Message: "refusing to clear an empty Redis key prefix"}
	}
	var cursor uint64
	pattern := string(append(clone(s.options.KeyPrefix), '*'))
	for {
		keys, next, err := s.client.Scan(ctx, cursor, pattern, 1024).Result()
		if err != nil {
			return redisError("clear_scan", err)
		}
		if len(keys) != 0 {
			if err := s.client.Del(ctx, keys...).Err(); err != nil {
				return redisError("clear_delete", err)
			}
		}
		if next == 0 {
			return nil
		}
		cursor = next
	}
}

func (s *Store) get(ctx context.Context, key string) (prolly.OptionalBytes, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.OptionalBytes{}, err
	}
	value, err := s.client.Get(ctx, key).Bytes()
	if errors.Is(err, redisclient.Nil) {
		return prolly.MissingBytes(), nil
	}
	if err != nil {
		return prolly.OptionalBytes{}, redisError("get", err)
	}
	return prolly.PresentBytes(value), nil
}
func (s *Store) familyKey(family, suffix []byte) string {
	value := make([]byte, 0, len(s.options.KeyPrefix)+len(family)+len(suffix))
	value = append(value, s.options.KeyPrefix...)
	value = append(value, family...)
	value = append(value, suffix...)
	return string(value)
}
func (s *Store) nodeKey(key []byte) string { return s.familyKey([]byte("node:"), key) }
func (s *Store) rootKey(key []byte) string { return s.familyKey([]byte("root:"), key) }
func (s *Store) hintKey(namespace, key []byte) string {
	suffix := make([]byte, 8, 8+len(namespace)+len(key))
	binary.BigEndian.PutUint64(suffix, uint64(len(namespace)))
	suffix = append(suffix, namespace...)
	suffix = append(suffix, key...)
	return s.familyKey([]byte("hint:"), suffix)
}
func (s *Store) scanFamily(ctx context.Context, family []byte) ([][]byte, error) {
	prefix := []byte(s.familyKey(family, nil))
	pattern := string(append(clone(prefix), '*'))
	var cursor uint64
	var result [][]byte
	for {
		keys, next, err := s.client.Scan(ctx, cursor, pattern, 1024).Result()
		if err != nil {
			return nil, redisError("scan", err)
		}
		for _, key := range keys {
			raw := []byte(key)
			if bytes.HasPrefix(raw, prefix) {
				result = append(result, clone(raw[len(prefix):]))
			}
		}
		if next == 0 {
			return result, nil
		}
		cursor = next
	}
}
func (s *Store) ready(ctx context.Context) error {
	if ctx != nil && ctx.Err() != nil {
		return ctx.Err()
	}
	if s == nil || s.client == nil {
		return &prolly.StoreError{Code: "invalid_store", Message: "Redis client is nil"}
	}
	return nil
}
func boolString(value bool) string {
	if value {
		return "1"
	}
	return "0"
}
func redisBool(value interface{}) (bool, error) {
	switch typed := value.(type) {
	case int64:
		return typed == 1, nil
	case bool:
		return typed, nil
	default:
		return false, invalidRedisResult("boolean", value)
	}
}
func redisOptional(value interface{}) (prolly.OptionalBytes, error) {
	switch typed := value.(type) {
	case nil:
		return prolly.MissingBytes(), nil
	case bool:
		if !typed {
			return prolly.MissingBytes(), nil
		}
	case string:
		return prolly.PresentBytes([]byte(typed)), nil
	case []byte:
		return prolly.PresentBytes(typed), nil
	}
	return prolly.OptionalBytes{}, invalidRedisResult("optional bytes", value)
}
func invalidRedisResult(context string, value interface{}) error {
	return &prolly.StoreError{Code: "invalid_result", Message: fmt.Sprintf("Redis %s returned %T", context, value)}
}
func redisError(operation string, err error) error {
	if err == nil {
		return nil
	}
	retryable := errors.Is(err, context.DeadlineExceeded) || strings.Contains(strings.ToLower(err.Error()), "connection")
	return &prolly.StoreError{Code: "redis", Message: fmt.Sprintf("%s: %v", operation, err), Retryable: retryable, Cause: err}
}
func clone(value []byte) []byte { return append([]byte(nil), value...) }

var _ prolly.RemoteStore = (*Store)(nil)

File: bindings/ruby/stores/redis/lib/prolly/store/redis.rb
# frozen_string_literal: true

require 'redis'
require 'prolly'

module Prolly
  class RedisRemoteStore < ForeignRemoteStore
    CAS_SCRIPT = <<~LUA.freeze
      local current = redis.call('GET', KEYS[1])
      local expected_present = ARGV[1] == '1'
      if expected_present then
        if current == false or current ~= ARGV[2] then
          return {0, current == false and 0 or 1, current or ''}
        end
      elseif current ~= false then
        return {0, 1, current}
      end
      if ARGV[3] == '1' then
        redis.call('SET', KEYS[1], ARGV[4])
        return {1, 1, ARGV[4]}
      end
      redis.call('DEL', KEYS[1])
      return {1, 0, ''}
    LUA

    MUTATE_SCRIPT = <<~LUA.freeze
      for index = 1, #KEYS do
        local offset = (index - 1) * 2
        if ARGV[offset + 1] == '1' then
          redis.call('SET', KEYS[index], ARGV[offset + 2])
        else
          redis.call('DEL', KEYS[index])
        end
      end
      return 1
    LUA

    TRANSACTION_SCRIPT = <<~LUA.freeze
      local condition_count = tonumber(ARGV[1])
      local node_count = tonumber(ARGV[2])
      local root_count = tonumber(ARGV[3])
      local argument = 4
      for index = 1, condition_count do
        local current = redis.call('GET', KEYS[index])
        local expected_present = ARGV[argument] == '1'
        local matches = (expected_present and current ~= false and current == ARGV[argument + 1])
          or (not expected_present and current == false)
        if not matches then
          return {0, index, current == false and 0 or 1, current or ''}
        end
        argument = argument + 2
      end
      local key_index = condition_count + 1
      for _ = 1, node_count do
        if ARGV[argument] == '1' then redis.call('SET', KEYS[key_index], ARGV[argument + 1])
        else redis.call('DEL', KEYS[key_index]) end
        argument = argument + 2
        key_index = key_index + 1
      end
      for _ = 1, root_count do
        if ARGV[argument] == '1' then redis.call('SET', KEYS[key_index], ARGV[argument + 1])
        else redis.call('DEL', KEYS[key_index]) end
        argument = argument + 2
        key_index = key_index + 1
      end
      return {1}
    LUA

    def initialize(client, key_prefix: 'prolly:'.b)
      raise ArgumentError, 'client must be Redis' unless client.is_a?(Redis)

      @client = client
      @prefix = key_prefix.b
      @mutex = Mutex.new
      @closed = false
    end

    def close = @closed = true

    def clear_namespace
      synchronize do
        raise ArgumentError, 'refusing to clear an empty Redis key prefix' if @prefix.empty?

        keys = scan_family(@prefix)
        keys.each_slice(256) { |slice| @client.del(*slice) }
      end
    end

    def descriptor
      value = StoreDescriptorRecord.new(
        protocol_major: 1, adapter_name: 'redis-v1', provider: 'redis', schema_version: 1,
        capabilities: StoreCapabilitiesRecord.new(
          native_batch_reads: true, atomic_batch_writes: true, node_scan: true,
          hints: true, atomic_nodes_and_hint: true, root_scan: true,
          root_compare_and_swap: true, transactions: true, read_parallelism: 1
        ),
        limits: StoreLimitsRecord.new(
          max_batch_read_items: nil, max_batch_write_items: nil,
          max_transaction_operations: nil, max_node_bytes: nil
        )
      )
      StoreDescriptorResultRecord.new(value: value, error: nil)
    end

    def get_node(cid) = optional_call { @client.get(family('node:', cid)) }
    def put_node(cid, value) = unit_call { @client.set(family('node:', cid), value.b) }
    def delete_node(cid) = unit_call { @client.del(family('node:', cid)) }

    def batch_nodes(operations)
      keys = operations.map { |item| family('node:', item.key) }
      arguments = operations.flat_map { |item| item.value.present ? ['1'.b, item.value.value.b] : ['0'.b, ''.b] }
      mutate(keys, arguments)
      unit
    rescue StandardError => error
      UnitResultRecord.new(error: store_error(error))
    end

    def batch_get_nodes_ordered(cids)
      values = synchronize do
        keys = cids.map { |cid| family('node:', cid) }
        keys.empty? ? [] : @client.mget(*keys).map { |value| optional(value) }
      end
      OptionalBytesListResultRecord.new(values: values, error: nil)
    rescue StandardError => error
      OptionalBytesListResultRecord.new(values: [], error: store_error(error))
    end

    def list_node_cids
      family_prefix = @prefix + 'node:'.b
      values = synchronize do
        scan_family(family_prefix).filter_map do |key|
          cid = key.byteslice(family_prefix.bytesize..)
          cid if cid.bytesize == 32
        end.sort
      end
      BytesListResultRecord.new(values: values, error: nil)
    rescue StandardError => error
      BytesListResultRecord.new(values: [], error: store_error(error))
    end

    def get_hint(namespace, hint_key) = optional_call { @client.get(hint_key(namespace, hint_key)) }
    def put_hint(namespace, hint_key, value) = unit_call { @client.set(hint_key(namespace, hint_key), value.b) }

    def batch_put_nodes_with_hint(nodes, namespace, hint_key_value, value)
      keys = nodes.map { |node| family('node:', node.key) }
      arguments = nodes.flat_map { |node| ['1'.b, node.value.b] }
      keys << hint_key(namespace, hint_key_value)
      arguments.concat(['1'.b, value.b])
      mutate(keys, arguments)
      unit
    rescue StandardError => error
      UnitResultRecord.new(error: store_error(error))
    end

    def get_root_manifest(name) = optional_call { @client.get(family('root:', name)) }
    def put_root_manifest(name, manifest) = unit_call { @client.set(family('root:', name), manifest.b) }
    def delete_root_manifest(name) = unit_call { @client.del(family('root:', name)) }

    def compare_and_swap_root_manifest(name, expected, replacement)
      response = synchronize do
        @client.eval(
          CAS_SCRIPT, keys: [family('root:', name)],
          argv: [flag(expected.present), expected.value.b, flag(replacement.present), replacement.value.b]
        )
      end
      validate_array(response, 3, 'CAS')
      RootCasResultRecord.new(
        applied: response[0].to_i == 1,
        current: optional_parts(response[1], response[2]), error: nil
      )
    rescue StandardError => error
      RootCasResultRecord.new(applied: false, current: optional(nil), error: store_error(error))
    end

    def list_root_manifests
      family_prefix = @prefix + 'root:'.b
      values = synchronize do
        keys = scan_family(family_prefix).sort
        manifests = keys.empty? ? [] : @client.mget(*keys)
        keys.zip(manifests).filter_map do |key, manifest|
          NamedBytesRecord.new(name: key.byteslice(family_prefix.bytesize..), value: manifest.b) if manifest
        end
      end
      NamedBytesListResultRecord.new(values: values, error: nil)
    rescue StandardError => error
      NamedBytesListResultRecord.new(values: [], error: store_error(error))
    end

    def commit_transaction(nodes, conditions, roots)
      keys = conditions.map { |item| family('root:', item.name) }
      keys.concat(nodes.map { |item| family('node:', item.key) })
      keys.concat(roots.map { |item| family('root:', item.name) })
      arguments = [conditions.length, nodes.length, roots.length]
      conditions.each { |item| arguments.concat([flag(item.expected.present), item.expected.value.b]) }
      nodes.each { |item| arguments.concat([flag(item.value.present), item.value.value.b]) }
      roots.each { |item| arguments.concat([flag(item.replacement.present), item.replacement.value.b]) }
      response = synchronize { @client.eval(TRANSACTION_SCRIPT, keys: keys, argv: arguments) }
      raise 'Redis returned an invalid transaction response' unless response.is_a?(Array) && !response.empty?
      return TransactionResultRecord.new(applied: true, conflict: nil, error: nil) if response[0].to_i == 1

      validate_array(response, 4, 'transaction conflict')
      index = response[1].to_i - 1
      raise 'Redis returned an invalid conflict index' unless index.between?(0, conditions.length - 1)
      condition = conditions[index]
      conflict = StoreTransactionConflictRecord.new(
        name: condition.name, expected: condition.expected,
        current: optional_parts(response[2], response[3])
      )
      TransactionResultRecord.new(applied: false, conflict: conflict, error: nil)
    rescue StandardError => error
      TransactionResultRecord.new(applied: false, conflict: nil, error: store_error(error))
    end

    private

    def synchronize(&block)
      raise 'Redis store is closed' if @closed
      @mutex.synchronize(&block)
    end

    def family(name, suffix) = @prefix + name.b + suffix.b
    def hint_key(namespace, key) = @prefix + 'hint:'.b + [namespace.bytesize].pack('Q>') + namespace.b + key.b
    def flag(value) = value ? '1'.b : '0'.b

    def scan_family(prefix)
      @client.scan_each.select { |key| key.b.start_with?(prefix) }.map(&:b)
    end

    def mutate(keys, arguments)
      synchronize { @client.eval(MUTATE_SCRIPT, keys: keys, argv: arguments) } unless keys.empty?
    end

    def optional(value) = OptionalBytesRecord.new(present: !value.nil?, value: value&.b || ''.b)
    def optional_parts(present, value) = present.to_i == 1 ? optional(value.b) : optional(nil)
    def unit = UnitResultRecord.new(error: nil)

    def optional_call
      OptionalBytesResultRecord.new(value: optional(synchronize { yield }), error: nil)
    rescue StandardError => error
      OptionalBytesResultRecord.new(value: optional(nil), error: store_error(error))
    end

    def unit_call
      synchronize { yield }
      unit
    rescue StandardError => error
      UnitResultRecord.new(error: store_error(error))
    end

    def validate_array(value, minimum, label)
      raise "Redis returned an invalid #{label} response" unless value.is_a?(Array) && value.length >= minimum
    end

    def store_error(error)
      StoreErrorRecord.new(
        code: error.is_a?(ArgumentError) ? 'invalid_argument' : 'internal',
        message: 'Redis provider operation failed', retryable: false,
        provider_code: error.class.name
      )
    end
  end
end
